mongo/document/document.go

```diff
@@ -0,0 +1,144 @@
+package document
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+
+	"github.com/Scalingo/go-utils/logger"
+	"github.com/Scalingo/go-utils/mongo"
+)
+
+type Document interface {
+	GetID() bson.ObjectId
+}
+
+type ParanoiaDeletable interface {
+	Updatable
+	SetDeletedAt(time.Time)
+}
+
+type Updatable interface {
+	Document
+	SetUpdatedAt(time.Time)
+}
+
+func Save(ctx context.Context, collectionName string, doc Document) error {
+	log := logger.Get(ctx)
+	c := mongo.Session(log).Clone().DB("").C(collectionName)
+	defer c.Database.Session.Close()
+	log.WithField(collectionName, doc).Debugf("save '%v'", collectionName)
+	_, err := c.UpsertId(doc.GetID(), &doc)
+	return err
+}
+
+// Remove the volume from the database.
+// Handle with care...
+func Destroy(ctx context.Context, collectionName string, doc Document) error {
+	log := logger.Get(ctx)
+	c := mongo.Session(log).Clone().DB("").C(collectionName)
+	defer c.Database.Session.Close()
+	log.WithField(collectionName, doc).Debugf("remove '%v'", collectionName)
+	return c.RemoveId(doc.GetID())
+}
+
+func ParanoiaDelete(ctx context.Context, collectionName string, d ParanoiaDeletable) error {
+	now := time.Now()
+	d.SetDeletedAt(now)
+	err := Update(ctx, collectionName, bson.M{"$set": bson.M{"deleted_at": now}}, d)
+	if err != nil {
+		return fmt.Errorf("fail to run mongo update: %v", err)
+	}
+	return nil
+}
+
+func Find(ctx context.Context, collectionName string, id bson.ObjectId, doc Document) error {
+	query := bson.M{"_id": id, "deleted_at": nil}
+	return FindOne(ctx, collectionName, query, doc)
+}
+
+func FindOne(ctx context.Context, collectionName string, query bson.M, doc Document) error {
+	log := logger.Get(ctx)
+	c := mongo.Session(log).Clone().DB("").C(collectionName)
+	defer c.Database.Session.Close()
+	return c.Find(query).One(doc)
+}
+
+func WhereParanoia(ctx context.Context, collectionName string, query bson.M, data interface{}) error {
```

Comment:
Don't you prefer to have a `Where` method which by default exclude documents with a defined `DeletedAt` field?

Reply:
Hmm, no I prefer that Where is doing.. where actually.